pkg/metric_aggregation/aggregation.go

```diff
@@ -0,0 +1,224 @@
+package aggregation
+
+import (
+	"fmt"
+	"io"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/kube-state-metrics/pkg/metric"
+	"strconv"
+	"strings"
+)
+
+// A metric value that will eventually be aggregated with others sharing the same LabelValues
+// Since the aggregation function can return multiple sets of label values,
+// a single Metric can correspond to multiple AggregatedValues.
+// Technically we could store the value just once and make LabelValues a [][]string
+// but keeping it this way makes the code simpler at minimal cost (a bunch of floats)
+type AggregatedValue struct {
+	Value       float64
+	LabelValues []string
+}
+
+// An Aggregation supported by a metric family generated by a FamilyGenerator
+// The Aggregate() function takes an object (same as passed to the generator's
+// GenerateFunc()) and the metric family as returned by the GenerateFunc().
+// It returns an array of AggregatedValues, each corresponding to a separate
+// metric in the aggregated family
+type Aggregation struct {
+	LabelNames []string
+	Aggregate  func(interface{}, metric.FamilyInterface) []AggregatedValue
+}
+
+// The runtime representation of an aggregated metric
+type AggregatedMetric struct {
+	// metric family metadata, generated once while creating the object
+	Header     string
+	Name       string
+	MetricType metric.Type
+
+	// AggregatedValues for each individual Kubernetes object
+	Objects map[types.UID][]AggregatedValue
+
+	// label names and the aggregation function come directly from the Aggregation object
+	LabelNames      []string
+	AggregationFunc func(interface{}, metric.FamilyInterface) []AggregatedValue
+}
+
+// AggregationSet represents all aggregations available from a single FamilyGenerator
+type AggregationSet struct {
+	aggregations map[string]*AggregatedMetric
+
+	// Aggregations are calculated lazily: an update to the stored metrics
+	// sets this flag to false and then the aggregates are recalculated
+	// during WriteAll().
```

Comment:
This is interesting -- the overhead for the aggregation calculation will be part of the latency of the polling request from the prometheus server, right? I don't know what other overhead is normally built into the handle for the /metrics request, but I guess in some extreme cases the calculation of aggregates could be expensive. Since the AggregationFunc is (apparently) a generic interface, we will need to be careful with what functions are allowed, to make sure they would not slow down metrics collection for the prom server. Is my reasoning correct here?

Reply:
The aggregation functions are called in response to cluster state updates, not during the `/metrics` request. The lazy part is adding all the values together.